### PR TITLE
support node label update

### LINKED ISSUE #24645: Docker 1.12 RC4 swarm mode - can't assign label to a node using docker node update command
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        
 OS/Arch:      linux/amd64
Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: docker-8:1-561338-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 89.33 MB
 Data Space Total: 107.4 GB
 Data Space Available: 9.149 GB
 Metadata Space Used: 692.2 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
Cgroup Driver: cgroupfs
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: active
 NodeID: cwi3c7ufsp9isz655oc8kpxpg
 IsManager: Yes
 Managers: 1
 Nodes: 2
 CACertHash: sha256:39e643b316da5d78c3fe32c506895fda2db36667efbf1614e8cb667b9db757f6
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.614 GiB
Name: managed-instance-4
ID: KH5Y:JLVS:YGQL:AACO:WXQJ:EQZ2:FY5M:OMHV:WWWG:VOBB:NCUZ:RXPM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VM on Google cloud

**Steps to reproduce the issue:**
1. Documentation (https://docs.docker.com/engine/reference/commandline/service_create/) says 

> Swarm administrators add node.labels for operational purposes by using the docker node update command.

2. Checking how to add label to a node 
```
$ sudo docker node update --help
Usage:  docker node update [OPTIONS] NODE
Update a node
Options:
      --availability string   Availability of the node (active/pause/drain)
      --help                  Print usage
      --membership string     Membership of the node (accepted/rejected)
      --role string           Role of the node (worker/manager)
```

**Describe the results you received:**
There is no option available to add/update label to a node

**Describe the results you expected:**
Have parameter to add label to node.labels context of the node using _docker node update_ command

**Additional information you deem important (e.g. issue happens only occasionally):**
Permanent

### PR DESCRIPTION
**Pull request by dongluochen**
**\- What I did**

Add CLI to update node labels. 

**\- How to verify it**

Use the following commands to validate the change. 

```
docker node update --label-add key1=value1 --label-add key2=value2 mynode
docker node inspect mynode
docker node update --label-rm key1 --label-rm key2 mynode
docker node inspect mynode
```

**\- Description for the changelog**
Support node label update

cc @diogomonica @aaronlehmann @aluzzardi .

closes https://github.com/docker/docker/issues/24645

### CODE REVIEW

**Comment by aaronlehmann** [comment]
Code LGTM, but docs need to be updated.

**Comment by thaJeztah** [comment]
+1 on design

**Comment by thaJeztah** [comment]
guess this is for 1.12?

**Comment by vdemeester** [comment]
@dongluochen @aaronlehmann @thaJeztah I think we should probably think of engine label too when update node labels. With this PR, we are going to introduce some inconsistency between the node label and the engine label (for the same node).

How swarm mode should *play*  with the daemon configuration on labels (not even talking about `--label` flag on `dockerd` but this is definitely gonna cause some trouble), and the reloading of restarting of the daemon ? How node labels update should deal with daemon configuration of the nodes ? How do we deal with *inconsistencies* between the label the node has in the cluster and the one the engine has when looked at alone ?

/cc @stevvooe @icecrime @aluzzardi @vieux 

(Puting back into `desgin-review` because I think this needs a little bit more discussion 👼)

**Comment by diogomonica** [comment]
@vdemeester I agree that this might be confusing, but the objective is to have two distinct types of labels: swarm labels and engine labels.

Engine labels can't be trusted for security sensitive scheduling decisions, since any worker can report any label up to a manager. They are, however, useful for things like "Does this host have an SSD?", or "how much ram/disk space does this host have?".

These swarm labels, on the other hand, are always trustworthy, since they are set by an administrator explicitly. This means that we can label a host as `pci-dss`, and create a constraint to ensure that `credit-card-app` only gets scheduled on worker nodes that are labeled as `pci-dss`, and that no malicious worker can self-label itself as `pci-dss`.

I hope this makes sense.

**Comment by thaJeztah** [comment]
Alright, so they're completely independent; should we deprecate the old labels, or only add a note to the docs?

**Comment by diogomonica** [comment]
I think they serve different purposes. it will always be useful to have a "self-reported" label coming from an engine. So I don't think we should deprecate engine labels.

Imagine that a host gets upgraded, and an SSD gets added. It is perfectly fine (and in fact, desired), for the host to update it's own label to include "I have an SSD". This is the function of the engine label.

However, for any label that never needs/shoudn't be updated by the node itself, swarm labels are the way to go.

**Comment by vdemeester** [comment]
@diogomonica This does make sence, thanks :+1:

**Comment by dongluochen** [comment]
Following is doc update on constraints from swarmkit [README](https://github.com/docker/swarmkit/blob/master/README.md). @thaJeztah Do you know which doc in docker/docker I should update? 

**Constraints**: Operators can limit the set of nodes where a task can be scheduled by defining constraint expressions. Multiple constraints find nodes that satisfy every expression, i.e., an `AND` match. Constraints can match node attributes in the following table. Note that `engine.labels` are collected from Docker Engine with information like operating system, drivers, etc. `node.labels` are added by cluster administrators for operational purpose. For example, some nodes have security compliant labels to run tasks with compliant requirements.


| node attribute | matches | example |
|:------------- |:-------------| :-------------|
| node.id | node's ID | `node.id == 2ivku8v2gvtg4`|
| node.hostname | node's hostname | `node.hostname != node-2`|
| node.role |  node's manager or worker role | `node.role == manager`|
| node.labels | node's labels added by cluster admins | `node.labels.security == high`|
| engine.labels | Docker Engine's labels | `engine.labels.operatingsystem == ubuntu 14.04`|

**Comment by thaJeztah** [comment]
@dongluochen I don't think it's properly documented yet; see https://github.com/docker/docker/issues/24104

/cc @sfsmithcha

**Comment by sfsmithcha** [comment]
@dongluochen I added your information above to the service_create.md See the following PR.

https://github.com/docker/docker/pull/24432

**Comment by sfsmithcha** [comment]
@dongluochen , added information about node update. See https://github.com/docker/docker/pull/24507.

**Comment by vdemeester** [comment]
Oh never actually did it…
Design LGTM, and given @thaJeztah @aaronlehmann so putting this in `code-review` 👼

**Comment by vdemeester** [comment]
Flags are consistent with #23773 :wink: 
Code LGTM too 🐸

File: api/client/node/update.go [hunk 1]
```diff
@@ -27,6 +28,8 @@ func newUpdateCommand(dockerCli *client.DockerCli) *cobra.Command {
 	flags.StringVar(&opts.role, flagRole, "", "Role of the node (worker/manager)")
 	flags.StringVar(&opts.membership, flagMembership, "", "Membership of the node (accepted/rejected)")
 	flags.StringVar(&opts.availability, flagAvailability, "", "Availability of the node (active/pause/drain)")
+	flags.StringSlice(flagLabelAdd, []string{}, "Add or update node labels (key1=value1,key2=value2)")
+	flags.StringSlice(flagLabelRemove, []string{}, "Remove node labels (key1,key2)")
```
Comment (thaJeztah) [inline api/client/node/update.go]: Noticed this while reviewing https://github.com/docker/docker/pull/24507#discussion_r70327976

Looks like we use a `StringSlice` here, which makes this flag support a comma-separated list. https://github.com/docker/docker/blob/28676fc04b9a307718b66bd28e8d9281d05f38ac/vendor/src/github.com/spf13/pflag/string_slice.go#L46-L54

We don't support this on other flags (and not on `--label`), so I don't think we should support that here

**Comment by thaJeztah** [comment]
I'm moving this back to design review, for https://github.com/docker/docker/pull/24397#discussion_r70338605

**Comment by thaJeztah** [comment]
ping @dongluochen PTAL ^^

**Comment by dongluochen** [comment]
Thanks @thaJeztah. I have a question about the format. I tried to follow https://github.com/docker/docker/pull/23773. But `--label-add` and `--label-rm` handling are different there.

**Comment by thaJeztah** [comment]
@dongluochen ah, yes, I saw your comment; I think we should not support it there as well (https://github.com/docker/docker/pull/23773#issuecomment-231919141). It's weird to add a new syntax/format on *some* flags, so better not have it (and possibly add it to all flags in a later stage if we really want it)

**Comment by dongluochen** [comment]
Change `label-add`/`label-rm` to update one label. PTAL. 

```
docker node update --label-add key1=value1 --label-add key2=value2 mynode
docker node update --label-rm key1 --label-rm key2 mynode
```

**Comment by vdemeester** [comment]
@thaJeztah putting it back to `code-review` then :wink: 
@dongluochen This will need some documentation updates (in `docs/references/commandline/node_update.md`at least)

**Comment by vdemeester** [comment]
@dongluochen needs a rebase too :sweat:

**Comment by dongluochen** [comment]
Update doc. Rebased. Thanks @vdemeester!

**Comment by thaJeztah** [comment]
Thanks! Found a couple of issues. (I'll check on the `service update` command as well if that has the same issues)

**1. Labels don't accept "no" value**

This is inconsistent with other labels; for containers, the `=value` is optional, and defaults to an empty string;
https://github.com/docker/docker/blob/v1.12.0-rc4/docs/reference/commandline/run.md#set-metadata-on-container--l---label---label-file

> The my-label key doesn't specify a value so the label defaults to an empty string("").


```bash
$ docker node update --label-add foo=bar --label-add blah 9a72qojtdvjvo59nkrloz1pl8
invalid argument "blah" for --label-add: bad attribute format: blah
See 'docker node update --help'.
```

Whereas this works;

```bash
$ docker run --label foo --name hello hello-world

$ docker inspect --format '{{json .Config.Labels }}' hello
{"foo":""}
```

**2. "label-rm" accepts "key=value"**

The `--label-rm` flag accepts `key=value`, and silently ignores the flag;

```bash
$ docker node update --label-add foo=bar 9a72qojtdvjvo59nkrloz1pl8
9a72qojtdvjvo59nkrloz1pl8

$ docker node inspect --format '{{ json .Spec.Labels}}' 9a72qojtdvjvo59nkrloz1pl8
{"foo":"bar"}
```

Using the "wrong" key/value:

```bash
$ docker node update --label-rm foo=hahaha 9a72qojtdvjvo59nkrloz1pl8
9a72qojtdvjvo59nkrloz1pl8

$ docker node inspect --format '{{ json .Spec.Labels}}' 9a72qojtdvjvo59nkrloz1pl8
{"foo":"bar"}

Using the "right" key/value:

```bash
$ docker node update --label-rm foo=bar 9a72qojtdvjvo59nkrloz1pl8
9a72qojtdvjvo59nkrloz1pl8

$ docker node inspect --format '{{ json .Spec.Labels}}' 9a72qojtdvjvo59nkrloz1pl8
{"foo":"bar"}
```

None of the above *remove* the label. We should either produce an error if `--label-rm` is a key/value pair, or "split" the `key` (but producing an error sounds like the better option here).

**3. Nodes without labels return `null`, not `{}`** 

Perhaps for a separate PR, but if no labels are set, `Spec.Labels` appears to be `null`, not an empty map;

```bash
docker node inspect --format '{{ json .Spec.Labels}}' 9a72qojtdvjvo59nkrloz1pl8
null
```

But for *containers*, it's always a map/object;

```bash
$ docker run --name nolabels busybox sh
$ docker inspect --format '{{json .Config.Labels }}' nolabels
{}
```

I think we should fix that. I'll open a separate issue

**Comment by thaJeztah** [comment]
actually; `3.` should probably be done in this PR as well, because this adds the "labels" feature (don't think we can add/remove labels on a node before this PR)

**Comment by thaJeztah** [comment]
Opened an issue for `service`, which has the same issue for empty labels; https://github.com/docker/docker/issues/24631

**Comment by dongluochen** [comment]
Thanks @thaJeztah for checking!

> 1 Labels don't accept "no" value

There are 2 problems here. The first one is engine labels require `opts.ValidateLabel`, i.e., `foo=bar` format. 
```
dchen@vm2:~/go/src/github.com/docker/docker$ grep -irnw . -e "ValidateLabel" --include=\*.go
./daemon/config.go:160:	cmd.Var(opts.NewNamedListOptsRef("labels", &config.Labels, opts.ValidateLabel), []string{"-label"}, usageFn("Set key=value labels to the daemon"))
./daemon/config.go:414:		if _, err := opts.ValidateLabel(label); err != nil {
``` 
The second problem is the meaning of a label with key but no value. Node labels are used by constraints (only for now). A label without value seems to indicate just the existence of a label. But current constraint doesn't support `exist` check.

> 2 "label-rm" accepts "key=value"

It takes the whole string "key=value" as key. It silently discards it because there is no such key. Does Docker support label remove in form of "key=value"? I prefer not to support it.  

> 3 Nodes without labels return null, not {}

I think this should be fixed at node join. An issue is opened at https://github.com/docker/swarmkit/issues/1160.

**Comment by thaJeztah** [comment]
> From what I see engine labels require opts.ValidateLabel, i.e., foo=bar format. So we are not consistent already.

The engine labels are the odd one out; all other labels (`container`, `image`, `volume`, `network`) support `--label foo` (without explicit value set), which defaults to `""` (empty string).

For that reason, I think we should support it, even if there's no use for it currently for constraints (people may have other uses for the labels)

> Does Docker support label remove in form of "key=value"? I prefer not to support it.

No, I don't think we should support it, since a label *key* cannot contain a `=`, I think it should produce an error.

> I think this should be fixed at node join

Not sure; if I do

```
docker node --label-add foo=bar <node>
docker node --label-rm foo <node>
```

I end up with `null` again, not `{}`. The point is that "no labels" results in `null`, and not a default empty `{}`

**Comment by dongluochen** [comment]
>> Does Docker support label remove in form of "key=value"? I prefer not to support it.

> No, I don't think we should support it, since a label key cannot contain a =, I think it should produce an error.

In the case of `--label-rm k1 --label-rm k2=v2`, should the command fail all of them, or fail only `k2=v2`? What should happen if a user tries to remove a non-existent label?

**Comment by thaJeztah** [comment]
If the wrong syntax is used, the entire command should fail (same as it currently fails when doing `--add-label=foo`).

> What should happen if a user tries to remove a non-existent label?

We can make the remove idempotent, after all, Swarm mode allows you to set a desired state; removing something that wasn't there is a no-op, and I don't see why we should produce an error in that case

>  Do we have example how such cases should be handled?

Hm, not sure, can't think of one right now, but ping me if I need to think harder :+1:

File: api/client/node/opts.go [hunk 2]
```diff
@@ -4,18 +4,37 @@ import (
...
 
 type nodeOptions struct {
+	annotations
 	role         string
 	membership   string
 	availability string
 }
 
+type annotations struct {
+	name   string
```
Comment (dnephin) [inline api/client/node/opts.go]: I think `name` is never set?

**Comment by dnephin** [comment]
LGTM

File: api/client/node/opts.go (hunk 2 above)
Comment (dongluochen) [inline api/client/node/opts.go]: Yes the name is never set. It's not clear how node.annotations.name should be used. So far we are using `hostname` only.

**Comment by dongluochen** [comment]
> removing something that wasn't there is a no-op, and I don't see why we should produce an error in that case

I find it more consistent to report error when `--label-rm` doesn't remove a label. If label `k1` doesn't exist, either `--label-rm k1=v1` or `--label-rm k1` would return error. PR updated for this.

**Comment by dongluochen** [comment]
> Nodes without labels return null, not {}

Current change doesn't fix this issue. I think it should be fixed in Swarmkit. Will look into it.

**Comment by thaJeztah** [comment]
> I find it more consistent to report error when --label-rm doesn't remove a label. 

my line of thought also was that `--label-add` adds or *updates* a label, so `--label-rm` would remove or (no-op) if not exist. But I'm ok with this.

I just tested the latest version, and LGTM

**Comment by thaJeztah** [comment]
@dongluochen can you squash your commits?

**Comment by dongluochen** [comment]
@thaJeztah Updated, thanks for validating!

**Comment by thaJeztah** [comment]
thanks!

**Pull request merged**